src/Model_Hydro/Hydro_Init_ByFunction_AssignData.cpp

```diff
@@ -186,31 +208,76 @@ void Hydro_Init_ByFunction_AssignData( const int lv )
       case INIT_MAG_BYVECPOT_FUNC : MHD_Init_BField_ByVecPot_Function( lv );   break;
       default                     : Aux_Error( ERROR_INFO, "unsupported OPT__INIT_BFIELD_BYVECPOT (%d) !!\n", OPT__INIT_BFIELD_BYVECPOT );
    }
-#  endif
 
+   real (*Ax)[ CUBE(PS1+1) ] = NULL;
+   real (*Ay)[ CUBE(PS1+1) ] = NULL;
+   real (*Az)[ CUBE(PS1+1) ] = NULL;
+
+   if ( UseVecPot )
+   {
+      Ax = new real [OMP_NT][ CUBE(PS1+1) ];
+      Ay = new real [OMP_NT][ CUBE(PS1+1) ];
+      Az = new real [OMP_NT][ CUBE(PS1+1) ];
+   }
+#  endif // #ifdef MHD
 
-#  pragma omp parallel for schedule( runtime ) num_threads( OMP_NThread )
+
+#  pragma omp parallel for schedule( runtime ) num_threads( OMP_NT )
    for (int PID=0; PID<amr->NPatchComma[lv][1]; PID++)
    {
 //    1. set the magnetic field
 #     ifdef MHD
 
+//    1-1. compute vector potential
+//         --> compute the entire patch at once to avoid redundant calculations
+#     ifdef OPENMP
+      const int TID = omp_get_thread_num();
+#     else
+      const int TID = 0;
+#     endif
+
+      real *AxTID = ( UseVecPot ) ? Ax[TID] : NULL;
+      real *AyTID = ( UseVecPot ) ? Ay[TID] : NULL;
+      real *AzTID = ( UseVecPot ) ? Az[TID] : NULL;
+
+      if ( UseVecPot )
+      {
+         int idx = 0;
+
+         for (int k=0; k<PS1+1; k++) {  const double z = amr->patch[0][lv][PID]->EdgeL[2] + k*dh;
+         for (int j=0; j<PS1+1; j++) {  const double y = amr->patch[0][lv][PID]->EdgeL[1] + j*dh;
+         for (int i=0; i<PS1+1; i++) {  const double x = amr->patch[0][lv][PID]->EdgeL[0] + i*dh;
+
+            AxTID[idx] = (real)0.0;
+            AyTID[idx] = (real)0.0;
+            AzTID[idx] = (real)0.0;
+
+            if ( i != PS1 )   AxTID[idx] = MHD_ResetByUser_VecPot_Ptr( x+dh_2, y,      z,      0.0, 0.0, lv, 'x', NULL );
+            if ( j != PS1 )   AyTID[idx] = MHD_ResetByUser_VecPot_Ptr( x,      y+dh_2, z,      0.0, 0.0, lv, 'y', NULL );
+            if ( k != PS1 )   AzTID[idx] = MHD_ResetByUser_VecPot_Ptr( x,      y,      z+dh_2, 0.0, 0.0, lv, 'z', NULL );
+
+            idx ++;
+         }}} // i,j,k
+      } // if ( UseVecPot )
+
+
       if ( !OPT__INIT_BFIELD_BYVECPOT ) {
```

Comment:
- The code for resetting the initial B field is within this `if` block, which will be executed only if the B field is not initialized from vector potential (`OPT__INIT_BFIELD_BYVECPOT = 0`). Maybe we need to mention that in the wiki, update the condition as `if ( !OPT__INIT_BFIELD_BYVECPOT  ||  ResetMag )`, or move the inner `if ( ResetMag )` block out.
- The code in lines 239-261 for initializing `A[x/y/z]TID` is used only when `ResetMag = True`. Maybe we can move it into the `if ( ResetMag )` block.

Reply:
@hfhsieh 
- Good catch! I've fixed it. Now we can reset B fields even with `OPT__INIT_BFIELD_BYVECPOT`.
- It should be fine since `UseVecPot` is true only when `ResetMag` is true. Note that for clarity I've renamed `UseVecPot` as `ResetMag_UseVecPot`.